File: src/AltV.Net/Elements/Entities/Blip.cs
using System;
using AltV.Net.Data;
using AltV.Net.Elements.Args;
using AltV.Net.Native;

namespace AltV.Net.Elements.Entities
{
    public class Blip : WorldObject, IBlip
    {
        public override Position Position
        {
            get
            {
                CheckExistence();
                var position = Position.Zero;
                AltNative.Blip.Blip_GetPosition(NativePointer, ref position);
                return position;
            }
            set
            {
                CheckExistence();
                AltNative.Blip.Blip_SetPosition(NativePointer, value);
            }
        }

        public override short Dimension
        {
            get
            {
                CheckExistence();
                return AltNative.Blip.Blip_GetDimension(NativePointer);
            }
            set
            {
                CheckExistence();
                AltNative.Blip.Blip_SetDimension(NativePointer, value);
            }
        }

        public override void SetMetaData(string key, object value)
        {
            CheckExistence();
            var mValue = MValue.CreateFromObject(value);
            AltNative.Blip.Blip_SetMetaData(NativePointer, key, ref mValue);
        }

        public override bool GetMetaData<T>(string key, out T result)
        {
            CheckExistence();
            var mValue = MValue.Nil;
            AltNative.Blip.Blip_GetMetaData(NativePointer, key, ref mValue);
            if (!(mValue.ToObject() is T cast))
            {
                result = default;
                return false;
            }

            result = cast;
            return true;
        }

        public bool IsGlobal
        {
            get
            {
                CheckExistence();
                return AltNative.Blip.Blip_IsGlobal(NativePointer);
            }
        }

        public bool IsAttached
        {
            get
            {
                CheckExistence();
                return AltNative.Blip.Blip_IsAttached(NativePointer);
            }
        }

        public IEntity AttachedTo
        {
            get
            {
                CheckExistence();
                var entityType = BaseObjectType.Undefined;
                var entityPointer = AltNative.Blip.Blip_AttachedTo(NativePointer, ref entityType);
                if (entityPointer == IntPtr.Zero) return null;
                return Alt.Module.BaseEntityPool.GetOrCreate(entityPointer, entityType, out var entity) ? entity : null;
            }
        }

        public byte BlipType
        {
            get
            {
                CheckExistence();
                return AltNative.Blip.Blip_GetType(NativePointer);
            }
        }

        public ushort Sprite
        {
            set
            {
                CheckExistence();
                AltNative.Blip.Blip_SetSprite(NativePointer, value);
            }
        }

        public byte Color
        {
            set
            {
                CheckExistence();
                AltNative.Blip.Blip_SetColor(NativePointer, value);
            }
        }

        public bool Route
        {
            set
            {
                CheckExistence();
                AltNative.Blip.Blip_SetRoute(NativePointer, value);
            }
        }

        public byte RouteColor
        {
            set
            {
                CheckExistence();
                AltNative.Blip.Blip_SetRouteColor(NativePointer, value);
            }
        }

        public Blip(IntPtr nativePointer) : base(nativePointer, BaseObjectType.Blip)
        {
        }
        
        public void Remove()
        {
            Alt.RemoveBlip(this);
        }
    }
}

File: clrhost/src/altv-c-api/blip.cpp
#include "blip.h"

// Entity

void Blip_GetPosition(alt::IBlip* blip, position_t &position) {
    auto blipPosition = blip->GetPosition();
    position.x = blipPosition.x;
    position.y = blipPosition.y;
    position.z = blipPosition.z;
}

void Blip_SetPosition(alt::IBlip* blip, alt::Position pos) {
    blip->SetPosition(pos);
}

int16_t Blip_GetDimension(alt::IBlip* blip) {
    return blip->GetDimension();
}

void Blip_SetDimension(alt::IBlip* blip, uint16_t dimension) {
    blip->SetDimension(dimension);
}

void Blip_GetMetaData(alt::IBlip* blip, const char* key, alt::MValue &val) {
    val = blip->GetMetaData(key);
}

void Blip_SetMetaData(alt::IBlip* blip, const char* key, alt::MValue* val) {
    blip->SetMetaData(key, *val);
}

// Blip

bool Blip_IsGlobal(alt::IBlip* blip) {
    return blip->IsGlobal();
}

bool Blip_IsAttached(alt::IBlip* blip) {
    return blip->IsAttached();
}

alt::IEntity* Blip_AttachedTo(alt::IBlip* blip, alt::IBaseObject::Type &type) {
    type = blip->AttachedTo()->GetType();
    return blip->AttachedTo();
}

uint8_t Blip_GetType(alt::IBlip* blip) {
    return (uint8_t) blip->GetBlipType();
}

void Blip_SetSprite(alt::IBlip* blip, uint16_t sprite) {
    blip->SetSprite(sprite);
}

void Blip_SetColor(alt::IBlip* blip, uint8_t color) {
    blip->SetColor(color);
}

void Blip_SetRoute(alt::IBlip* blip, bool state) {
    blip->SetRoute(state);
}

void Blip_SetRouteColor(alt::IBlip* blip, uint8_t color) {
    blip->SetRouteColor(color);
}
